urlconnection/src/main/java/zipkin/reporter/urlconnection/URLConnectionReporter.java

```diff
@@ -0,0 +1,127 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin.reporter.urlconnection;
+
+import com.google.auto.value.AutoValue;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.zip.GZIPOutputStream;
+import zipkin.Codec;
+import zipkin.Span;
+import zipkin.storage.AsyncSpanConsumer;
+import zipkin.storage.Callback;
+
+/**
+ * Reports spans to Zipkin, using its {@code POST /spans} endpoint.
+ */
+@AutoValue
+public abstract class URLConnectionReporter implements AsyncSpanConsumer {
+
+  public static Builder builder() {
+    return new AutoValue_URLConnectionReporter.Builder()
+        .baseUrl("http://localhost:9411")
+        .connectTimeout(10 * 1000)
+        .readTimeout(60 * 1000)
+        .compressionEnabled(true)
+        .executor(Runnable::run);
+  }
+
+  @AutoValue.Builder
+  public interface Builder {
+    /** Default http://localhost:9411. URL of the zipkin server service. */
+    Builder baseUrl(String baseUrl);
+
+    /** Default 10 * 1000 milliseconds. 0 implies no timeout. */
+    Builder connectTimeout(int connectTimeout);
+
+    /** Default 60 * 1000 milliseconds. 0 implies no timeout. */
+    Builder readTimeout(int readTimeout);
+
+    /** Default true. true implies that spans will be gzipped before transport. */
+    Builder compressionEnabled(boolean compressSpans);
+
+    /** Default calling thread. The executor used to defer http requests. */
+    Builder executor(Executor executor);
+
+    URLConnectionReporter build();
+  }
+
+  public Builder toBuilder() {
+    return new AutoValue_URLConnectionReporter.Builder(this);
+  }
+
+  abstract String baseUrl();
+
+  abstract int connectTimeout();
+
+  abstract int readTimeout();
+
+  abstract boolean compressionEnabled();
+
+  abstract Executor executor();
+
+  @Override public void accept(List<Span> spans, Callback<Void> callback) {
+    executor().execute(() -> {
+      try {
+        send(spans);
+        callback.onSuccess(null);
+      } catch (RuntimeException | IOException | Error e) {
+        callback.onError(e);
+        if (e instanceof Error) throw (Error) e;
+      }
+    });
+  }
+
+  void send(List<Span> spans) throws IOException {
+    byte[] body = Codec.JSON.writeSpans(spans);
+    String mediaType = "application/json";
+    URL url = new URL(baseUrl() + (baseUrl().endsWith("/") ? "" : "/") + "api/v1/spans");
```

Comment:
Good point. Will make it possible to affect dynamically.

On 11 Aug 2016 6:17 am, "Nic Munroe" <notifications@github.com> wrote:

> In urlconnection/src/main/java/zipkin/reporter/urlconnection/
> URLConnectionReporter.java
> <https://github.com/openzipkin/zipkin-java-reporter/pull/2#discussion_r74341626>
> :
>
> > +  @Override public void accept(List<Span> spans, Callback<Void> callback) {
> > +    executor().execute(() -> {
> > +      try {
> > +        send(spans);
> > +        callback.onSuccess(null);
> > +      } catch (RuntimeException | IOException | Error e) {
> > +        callback.onError(e);
> > +        if (e instanceof Error) throw (Error) e;
> > +      }
> > +    });
> > +  }
> > +
> > +  void send(List<Span> spans) throws IOException {
> > +    byte[] body = Codec.JSON.writeSpans(spans);
> > +    String mediaType = "application/json";
> > +    URL url = new URL(baseUrl() + (baseUrl().endsWith("/") ? "" : "/") + "api/v1/spans");
>
> Is there a use case where the baseUrl could change during runtime? Why not
> create the URL once during app construction and reuse it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-java-reporter/pull/2/files/7ef109b5d90aae801125892a5a0ea26c248a8aa6#r74341626>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD6156USMuFed-LFxrFlypQhCu9X2kuks5qek3lgaJpZM4Jg-e3>
> .
>